gossip/src/push_active_set.rs

```diff
@@ -77,11 +105,54 @@ impl PushActiveSet {
             .iter()
             .map(|node| get_stake_bucket(stakes.get(node)))
             .collect();
+
+        // Get fraction of unstaked nodes in the cluster.
+        let num_unstaked = buckets.iter().filter(|&&b| b == 0).count();
+        let fraction_unstaked = num_unstaked as f64 / nodes.len() as f64;
+        let alpha_target = 1.0 + fraction_unstaked.min(1.0);
+
+        // SAFETY & CORRECTNESS NOTES
+        //
+        // `alpha` is updated with a first-order low-pass filter:
+        //
+        //     `alpha` = K * `alpha_target` + (1 – K) * `alpha_previous`
+        //
+        // where  0 < K < 1  (fixed at compile time, K ~ 0.611) and
+        //       `alpha_target` ∈ [1.0, 2.0]
+        //
+        // Why this cannot panic or go out of bounds:
+        //
+        // 1. **Finite/Guaranteed range** – All terms are `f64` values in
+        //    [1, 2]-> `K`, `alpha_target`, and `alpha_previous`.  Multiplication
+        //    and addition in this range cannot overflow an IEEE-754 `f64`
+        //    (overflow requires ~1e308).
+        //
+        // 2. **No division** – No divide-by-zero risk.
+        //
+        // 3. **NaN / +-Inf protection** – `contains` ensures any accidental `NaN` or
+        //    `+-Inf` (impossible here unless prior state is corrupt) reduces to DEFAULT_ALPHA.
+        //
+        // 4. **Lag / Overshoot protection** – `alpha_target` changes smoothly in practice
+        //    (small d(alpha_target)/dt). If future code changes make `alpha_target` jump
+        //    larger, we must retune `TC`/`K` or use a higher‑order filter to avoid
+        //    lag/overshoot.
+        let alpha = K * alpha_target + (1.0 - K) * self.alpha;
```

Comment:
Should we extract the IIR filter into its own module and stick it in net-utils or something like it? Would make it far easier to rewrite it into u64 and test that it does not overflow for no good reason.